gateway-ha/src/main/java/io/trino/gateway/ha/handler/ProxyUtils.java

```diff
@@ -89,34 +93,34 @@ public static String getQueryUser(String userHeader, String authorization)
         return parts.get(0);
     }
 
-    public static String extractQueryIdIfPresent(HttpServletRequest request, List<String> statementPaths)
+    public static String extractQueryIdIfPresent(
+            HttpServletRequest request,
+            List<String> statementPaths,
+            boolean requestAnalyserClientsUseV2Format,
+            int requestAnalyserMaxBodySize)
     {
         String path = request.getRequestURI();
         String queryParams = request.getQueryString();
+        if (!request.getMethod().equals(HttpMethod.POST)) {
+            return extractQueryIdIfPresent(path, queryParams, statementPaths);
+        }
+        String queryText = null;
         try {
-            String queryText = CharStreams.toString(new InputStreamReader(request.getInputStream()));
-            if (!isNullOrEmpty(queryText)
-                    && queryText.toLowerCase().contains("system.runtime.kill_query")) {
-                // extract and return the queryId
-                String[] parts = queryText.split(",");
-                for (String part : parts) {
-                    if (part.contains("query_id")) {
-                        Matcher matcher = EXTRACT_BETWEEN_SINGLE_QUOTES.matcher(part);
-                        if (matcher.find()) {
-                            String queryQuoted = matcher.group();
-                            if (!isNullOrEmpty(queryQuoted) && queryQuoted.length() > 0) {
-                                return queryQuoted.substring(1, queryQuoted.length() - 1);
-                            }
-                        }
-                    }
-                }
-            }
+            queryText = CharStreams.toString(new InputStreamReader(request.getInputStream()));
         }
-        catch (Exception e) {
-            log.error(e, "Error extracting query payload from request");
+        catch (IOException e) {
+            log.error(e, "Error reading request body");
+            throw new RuntimeException("Error reading request body", e);
         }
-
-        return extractQueryIdIfPresent(path, queryParams, statementPaths);
+        if (!isNullOrEmpty(queryText) && queryText.toLowerCase().contains("kill_query")) {
+            TrinoQueryProperties trinoQueryProperties = new TrinoQueryProperties(request, requestAnalyserClientsUseV2Format, requestAnalyserMaxBodySize);
+            if (trinoQueryProperties.getProcedure().filter(p -> p.getName().getSuffix().equals("kill_query")).isPresent()) {
+                checkArgument(trinoQueryProperties.getProcedureArguments().getFirst().getValue() instanceof StringLiteral,
+                        "Unable to route kill_query procedures where the first argument is not a String Literal");
+                return ((StringLiteral) trinoQueryProperties.getProcedureArguments().getFirst().getValue()).getValue();
+            }
+        }
+        return null;
```

Comment:
When do we reach this condition? Is it tested?

Reply:
Please add `@Nullable` annotation to this method.

Reply:
Unfortunately `null` is used to signify that no query ID was found in either the path, query parameters or request body. Any new query that is not a `kill_query`, or a GET to a path such as `/oauth2` or `/v1/info` will reach this condition. So it is tested by the routing tests like `TestGatewayHaMultipleBackends`.

That said, I will upgrade this logic to use `Optional`, as we have done in a few other places in the code, and I'll do a follow up PR to remove the use of `null` as a special value in other places as well